core/federated/RTI/enclave.c

```diff
@@ -0,0 +1,297 @@
+#include "enclave.h"
+
+/**
+ * Reference to enclave_RTI_t instance.
+ */
+enclave_RTI_t* _E_RTI;
+
+// Global variables defined in tag.c:
+extern instant_t start_time;
+
+// RTI mutex, which is the main lock  
+extern lf_mutex_t rti_mutex;
+
+// FIXME: rename "federate" everywhere in this file.
+// FIXME: Done, except for log and debug message. What sould be kept: 'enclave', 
+//        'federate', or 'enlcave/federate'?
+// FIXME: Should enclaves tracing use the same mechanism as federates? 
+//        It needs to account a federate having itself a number of enclaves.
+//        Currently, all calls to tracepoint_from_federate() and 
+//        tracepoint_to_federate() are in rti_lib.c
+
+void initialize_enclave(enclave_t* e, uint16_t id) {
+    e->id = id;
+    e->completed = NEVER_TAG;
+    e->last_granted = NEVER_TAG;
+    e->last_provisionally_granted = NEVER_TAG;
+    e->next_event = NEVER_TAG;
+    e->state = NOT_CONNECTED;
+    e->upstream = NULL;
+    e->upstream_delay = NULL;
+    e->num_upstream = 0;
+    e->downstream = NULL;
+    e->num_downstream = 0;
+    e->mode = REALTIME;
+    e->requested_stop = false;
+
+    // Initialize the next event condition variable.
+    lf_cond_init(&e->next_event_condition, &rti_mutex);
+}
+
+void logical_tag_complete(enclave_t* enclave, tag_t completed) {
+    // FIXME: Consolidate this message with NET to get NMR (Next Message Request).
+    // Careful with handling startup and shutdown.
+    lf_mutex_lock(&rti_mutex);
+
+    enclave->completed = completed;
+
+    LF_PRINT_LOG("RTI received from federate/enclave %d the Logical Tag Complete (LTC) " PRINTF_TAG ".",
+                enclave->id, enclave->completed.time - start_time, enclave->completed.microstep);
+
+    // See if we can remove any of the recorded in-transit messages for this.
+    // FIXME: Should this be here?
+    // clean_in_transit_message_record_up_to_tag(enclave->in_transit_message_tags, enclave->completed);
```

Comment:
This is commented out. But surely it is needed for the federates?

Reply:
Indeed, it is already called in `federate.c`.
To be removed.